aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -563,6 +564,68 @@ Tensor masked_fill(const Tensor & self, const Tensor & mask, const Tensor & sour
   return result;
 }
 
+static Tensor & masked_select_out_impl_cpu(Tensor & result, const Tensor & self, const Tensor & mask) {
+  NoNamesGuard guard;
+
+  TORCH_CHECK(mask.scalar_type() == ScalarType::Byte || mask.scalar_type() == ScalarType::Bool,
+              "masked_select: expected BoolTensor or ByteTensor for mask");
+  TORCH_CHECK(self.scalar_type() == result.scalar_type(),
+              "masked_select(): self and result must have the same scalar type");
+
+  if (mask.dtype() == at::ScalarType::Byte) {
+    AT_WARN("masked_select received a mask with dtype torch.uint8, this behavior is now deprecated," \
+            "please use a mask with dtype torch.bool instead.");
+  }
+
+  Tensor _mask, _self;
+  std::tie(_mask, _self) = expand_outplace(mask, self);
+
+  auto shape = _self.sizes().vec();
+  auto mask_long = at::empty(shape, self.options().dtype(at::kLong)).copy_(_mask);
+  int64_t numel = mask_long.sum().item().toLong();
```

Comment:
While working on my cuda version of masked_select, I realized that we can reuse the final element from the cumulative sum result to get `numel`, rather than running a whole other sum calculation.